src/kernels/composable_kernel/include/utility/reduction_operator.hpp

```diff
@@ -0,0 +1,177 @@
+/*******************************************************************************
+ *
+ * MIT License
+ *
+ * Copyright (c) 2020 Advanced Micro Devices, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+#ifndef CK_REDUCTION_OPERATOR_HPP
+#define CK_REDUCTION_OPERATOR_HPP
+
+#include <limits>
+#include "reduction_common.hpp"
+
+namespace ck {
+
+namespace reduce {
+
+template <class T>
+struct Add
+{
+    using dataType = T;
+
+    __device__ static T GetZeroVal() { return type_convert<T>{}(0.0f); };
+
+    __device__ inline constexpr void operator()(T& a, T b) const { a = a + b; }
+
+    static constexpr bool indexable = false;
+};
+
+template <class T>
+struct Mul
+{
+    using dataType = T;
+
+    __device__ static T GetZeroVal() { return type_convert<T>{}(1.0f); };
```

Comment:
I think it's better to call it ```GetInitialValue``` instead of ```GetZeroVal```, because it's initial value for reduction ```Mul```, not ```zero```` value of ```T```

Reply:
I prefer to use "zero value" or "unit value",  because it is from the set theory, and this value uniquely depends on the type of the binary operation defined by the space rather than depending on any context. 

I prefer to keep the name "GetZeroVal()" here and re-name to  "initial value"   when used by the Run() method of ThreadwiseGenericSliceCopy/BlockwiseGenericSlicCopy.

Reply:
That's fine. You don't need to change here or when you call Run().

But please add some comment here what "zero value" here means.

Reply:
[Recommendation] I see an edge between two realms here. Do whatever you think is better, but _please_ add respective comment.

Reply:
> [Recommendation] I see an edge between two realms here. Do whatever you think is better, but _please_ add respective comment.

Yep, some comment would be nice.

Reply:
[Notice] This piece of code is confusing, so if I were a reviewer, I would say that comment is a must here.

Reply:
Ping